### DISCUSSION TITLE
Unreserved keywords?

### DISCUSSION THREAD

**Discussion by ZedThree** [tree-sitter/tree-sitter#2084] [Q&A] [answered]
I'm trying to deal with Fortran's unreserved keywords -- it's perfectly fine to have `if` as a variable, for example:

```fortran
program foo
  integer :: if
  if = 1
  if (if == 1) then
    if = 2
  endif
end program foo
```

Tree-sitter doesn't like the assignments, but the `if` statement is fine:

```diff
  (translation_unit
    (program
      (program_statement (name))
      (variable_declaration (intrinsic_type) (identifier))
-      (ERROR)
-      (statement_label)
+     (assignment_statement (identifier) (number_literal))
       (if_statement
         (parenthesized_expression
           (relational_expression (identifier) (number_literal)))
-       (ERROR)
+       (assignment_statement (identifier) (number_literal))
        (end_if_statement))
      (end_program_statement (name))))
```

I've tried adding `word: $ => $.identifier` as described in the Keyword Extraction docs, but it made no difference. The `assignment_statement` already has `prec.right(-10, ...)`.

Any ideas or pointers on how to handle this? Fortran has a fair few keywords like this, including some like `data` and `save` that quite often clash with variable names.

**Comment by ahelwer** [comment]
Did you look at lexical vs. parse precedence

**Comment by sogaiu** [comment]
IIUC @ahelwer might be referring to what's discussed [here](https://tree-sitter.github.io/tree-sitter/creating-parsers#lexical-precedence-vs-parse-precedence).

**Comment by ZedThree** [comment]
I have, but I'm afraid I didn't really understand it enough to help. It reads as though the only thing we can do to fix it is to use an external scanner -- but then I'd be worried we'd end up having to use that for most tokens.

Here's a minimal grammar and example:

```js
module.exports = grammar({
  name: 'mvce',

  rules: {
    source_file: $ => repeat(
      choice(
        $.assignment_statement,
        $.if_statement
      )
    ),

    assignment_statement: $ => seq(
      $.identifier, '=', $.number_literal, '\n'
    ),

    if_statement: $ => seq(
      'if', '(', $.identifier, '==', $.number_literal, ')', '\n',
      $.assignment_statement,
      'end', 'if', '\n'
    ),

    number_literal: $ => /\d+/,

    identifier: $ => /[a-zA-Z_]\w*/,
  }
});
```
Example:

```
if = 1
if (hello == 2)
  goodbye = 3
end if
```

which gives:

```
(source_file [0, 0] - [4, 0]
  (ERROR [0, 0] - [0, 6]
    (number_literal [0, 5] - [0, 6]))
  (if_statement [1, 0] - [4, 0]
    (identifier [1, 4] - [1, 9])
    (number_literal [1, 13] - [1, 14])
    (assignment_statement [2, 2] - [3, 0]
      (identifier [2, 2] - [2, 9])
      (number_literal [2, 12] - [2, 13]))))
```

Any ideas on how to make it work with this minimal example?

**Reply by ahelwer** [reply]
It often helps to look at the parser's debug output with `-d`. You don't need to use an external scanner for this, did you see the section about using precedence in the `token` function? It says to pay particular attention to it.

**Reply by ZedThree** [reply]
The debug output is not particularly enlightening. Here it is up to the first `detect_error`:

```
process version:0, version_count:1, state:1, row:0, col:0
lex_internal state:7, row:0, column:0
  consume character:'i'
  consume character:'f'
lexed_lookahead sym:if, size:2
shift state:12
process version:0, version_count:1, state:12, row:0, col:2
lex_internal state:0, row:0, column:2
  skip character:' '
  consume character:'='
lexed_lookahead sym:=, size:2
detect_error
```

It consumes `if` and then immediately decides it's an `if_statement`. It errors when it comes to the `=`, and then resumes immediately after that. It's not clear what can be done with that information.

I'm very sorry for being so stupid, but I don't really understand how `token` will help. I'm not using `token` here, and I can't see where I would want to use it -- it produces leaf nodes, right? I can imagine that `if (` should perhaps be a single token (and `end if` too), but that doesn't seem to help.

I can also wrap the `identifier` regex in `token(prec(1, ...))` which causes `if` to always be parsed as an identifier and `if` statements to be errors.

Am I missing something obvious?

**Reply by ahelwer** [reply, marked as answer]
Hmmm, okay. I played around with it and was able to fix it by changing your identifier rule to:
```js
identifier: $ => choice(/[a-zA-Z_]\w*/, 'if')
```
In general you should add all your keywords as literals of the identifier in that way, if you're operating in a language where that is valid syntax. This means the lexer doesn't have to "commit" to labeling the string `if` as one of an `if` literal token or an `identifier` token, and can put off that categorization until seeing what comes after it.

**Reply by ZedThree** [reply]
Amazing, thanks! I see, because of the "Match Specificity" rule it wants to match to the plain string token rather than the regex.

It looks like we might also have to add conflicts in the real parser for each keyword, but it does seem to work, thanks!

**Reply by kennyhml** [reply]
@ahelwer Sorry to revive this discussion after so long, but is this currently still the only way?
I'm parsing ABAP, the language has hundreds of keywords - none of them are reserved. I came to this solution on my own, but with the sheer number of keyword literals in the choice, at some point the parser just gives up and starts running out of memory.
